### DISCUSSION TITLE
Unable to resolve service for type 'Elsa.Persistence.EntityFramework.Core.ElsaContext' while attempting to activate 'Controllers.BinApprovalController

### DISCUSSION THREAD

**Discussion by kajasumanie** [elsa-workflows/elsa-core#5413] [Design & tech]
I am working on an Elsa .NET 8 proof of concept (POC). I'm attempting to connect Elsacontext in my controller, but I'm encountering the following issue.

Unable to resolve service for type 'Elsa.Persistence.EntityFramework.Core.ElsaContext' while attempting to activate 'ElsaDemo2.Controllers.BinApprovalController

My program.cs code:-
--------------------------------------------------------------------

builder.Services.AddDbContext<ElsaContext>(options =>
{
    options.UseSqlite("Data Source=elsa.db;");
});

builder.Services.AddElsaCore(options =>
    options
        .UseEntityFrameworkPersistence(ef => ef.UseSqlite("Data Source=elsa.db;"), true)
        .AddHttpActivities()
        .AddWorkflow<HelloWorld>())
    .AddElsaApiEndpoints();

----------------------------------------------------------
My controller code:-
--------------------------------------------------------------

        [HttpPost("v1/bins")]
        public async Task<IActionResult> Post([FromBody] BinRequestModel request, CancellationToken cancellationToken)
        {
            var allVersions = await _elsaContext.WorkflowDefinitions.ToListAsync(cancellationToken);
            var latestVersion = allVersions.OrderByDescending(x => x.Version).FirstOrDefault();

            return Ok("<h1>Request for Approval Sent</h1><p>Your bin has been received and will be reviewed shortly.</p>");
        }

        public class BinRequestModel
        {
            public string BinId { get; set; }
            // Add additional properties as needed
        }

------------------------------------------------------------------------------------------------------
What I'm trying to do is test the Elsa database data using a POST request through a Postman collection.

Here is the URL and body I used for the Postman request:

https://localhost:xxxxx/api/BinApproval/v1/bins/

{ "BinId": "bin123" }

@sfmskywalker  Since I DI the context I am not sure what still wrong.

**Comment by glime-ai** [comment]
Based on the information provided, it seems that you are trying to inject `ElsaContext` directly into your controller, which is not resolving correctly. The `ElsaContext` is typically not registered directly with the DI container; instead, Elsa provides extension methods to set up the required services.

To resolve the issue, you should not inject `ElsaContext` directly. Instead, you should use the `ElsaContext` within a scope as it is a scoped service. You can achieve this by using the `IServiceScopeFactory` to create a scope and then get the `ElsaContext` from there.

Here's an example of how you can modify your controller to use `IServiceScopeFactory`:

```csharp
using Elsa.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.DependencyInjection;

namespace ElsaDemo2.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class BinApprovalController : ControllerBase
    {
        private readonly IWorkflowRegistry _workflowRegistry;
        private readonly IWorkflowLaunchpad _workflowLaunchpad;
        private readonly IServiceScopeFactory _scopeFactory;

        public BinApprovalController(IWorkflowRegistry workflowRegistry, IWorkflowLaunchpad workflowLaunchpad, IServiceScopeFactory scopeFactory)
        {
            _workflowRegistry = workflowRegistry;
            _workflowLaunchpad = workflowLaunchpad;
            _scopeFactory = scopeFactory;
        }

        [HttpPost("v1/bins")]
        public async Task<IActionResult> Post([FromBody] BinRequestModel request, CancellationToken cancellationToken)
        {
            using (var scope = _scopeFactory.CreateScope())
            {
                var elsaContext = scope.ServiceProvider.GetRequiredService<ElsaContext>();
                var allVersions = await elsaContext.WorkflowDefinitions.ToListAsync(cancellationToken);
                var latestVersion = allVersions.OrderByDescending(x => x.Version).FirstOrDefault();
            }

            return Ok("<h1>Request for Approval Sent</h1><p>Your bin has been received and will be reviewed shortly.</p>");
        }

        public class BinRequestModel
        {
            public string BinId { get; set; }
            // Add additional properties as needed
        }
    }
}
```

In this example, `IServiceScopeFactory` is injected into the controller, and a new scope is created within the `Post` method to resolve `ElsaContext`. This should resolve the issue you are facing with dependency injection.

Remember to remove the direct injection of `ElsaContext` from your controller's constructor, as it is not necessary and will not work as expected.

If you continue to face issues, please ensure that your Elsa packages are compatible with .NET 8 and that you are following the latest documentation for Elsa 3.x, as there might be differences from previous versions.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by kajasumanie** [reply]
Nope In line  var elsaContext = scope.ServiceProvider.GetRequiredService<ElsaContext>();, again the same issue I am facing

**Comment by sfmskywalker** [comment]
Elsa doesn’t register the ElsaContext, but rather a DbContextFactory. Try injecting something along the lines of: `IDbContextFactory<ElsaContext>`. Then use that factory in your controller actions and or filters to create an ElsaContext.

**Reply by kajasumanie** [reply]
@sfmskywalker Worked. Thanks
